### DISCUSSION TITLE
How do you serve static files properly.

### DISCUSSION THREAD

**Discussion by TwinFuture** [uNetworking/uWebSockets.js#110] [Q&A] [answered]
How would I go about serving static files such as images possibly with cache?. Is this safe to build a web application with uws? I also wanna say the performance on this is absolutely amazing. I will definitely be sharing my experience when finished.

This is just super super fast, with just res.end('hey'), on an i7 laptop I'm able to get 1 million req's a second! off the BAT! Data.

Stat           │ 1%         │ 2.5%      │ 50%       │ 97.5%   │ Avg            │ Stdev       │ Min       │
Req/Sec    │ 44991    │ 69759    │ 121087  │ 139263 │ 117386.57 │ 18901.81 │ 44972    │ Bytes/Sec │ 2.75 MB │ 4.26 MB │ 7.39 MB │ 8.5 MB │ 7.16 MB     │ 1.15 MB  │ 2.74 MB │

Req/Bytes counts sampled once per second.

6221k requests in 81.45s, 379 MB read
300 errors (300 timeouts) ONLY!!!

I had to stress test this using 8 clusters all firing and producing equal results to the ones posted. The actual test used 100% cpu on a single core for the stress test alone, so I used 8 cores to stress the server resulting in 8 different results.  The stress test was fired from a different machine. As you can see I hit 1 million req/s. I have basically gone from 250k req's to 1 million. wow. That's compared to nodes http server. Ooo and with only 25% cpu usage, compared to 100% from node. Also the only reason I am not able to max this cpu usage out is because to send that many requests maxed out the cpu on my desktop i7.

When serving a file say an html file using nginx I am getting very similar results to uws using. This is what I come up with to test with uws..
```
    // Read file from file system.
    let file = fs.readFileSync('index.html', function (err, data) {
        if (err) {
            res.end(`Error getting the file: ${err}.`);
        } else {
            //res.writeHeader('Content-type', 'text/html');
            res.end(data);
        }
    });
    res.end(file);
```
Which is the same for both nginx and uws for serving a file. Which is pretty awesome. These are some big numbers for a dinky i7 laptop that's pretty old now.

Sorry for the long post. Very appreciated.

**Comment by ghost** [comment, marked as answer]
There are examples such as VideoStreamer.js showing how to send potentially large files. Res.end is only for small responses

**Comment by PeterHewat** [comment]
Hi Alex, can you quantify "small responses" please?

**Comment by ghost** [comment]
Less than big ;)

**Comment by ghost** [comment]
Less than something you would care about if it wound up as backpressure

**Comment by PeterHewat** [comment]
So I guess up to 66.6KB is fine :o) But what about 666KB :-O
